src/java.base/share/classes/java/lang/foreign/SymbolLookup.java

```diff
@@ -0,0 +1,178 @@
+     * The returned lookup returns native symbols backed by a non-closeable, shared scope which keeps the caller's classloader
+     * <a href="../../../java/lang/ref/package.html#reachability">reachable</a>.
+     *
+     * @return a symbol lookup suitable to find symbols in libraries loaded by the caller's classloader.
+     * @see System#load(String)
+     * @see System#loadLibrary(String)
+     */
+    @CallerSensitive
+    static SymbolLookup loaderLookup() {
+        Class<?> caller = Reflection.getCallerClass();
+        ClassLoader loader = Objects.requireNonNull(caller.getClassLoader());
+        MemorySessionImpl loaderSession = MemorySessionImpl.heapSession(loader);
+        return name -> {
+            Objects.requireNonNull(name);
+            JavaLangAccess javaLangAccess = SharedSecrets.getJavaLangAccess();
+            MemoryAddress addr = MemoryAddress.ofLong(javaLangAccess.findNative(loader, name));
+            return addr == MemoryAddress.NULL
+                    ? Optional.empty() :
+                    Optional.of(MemorySegment.ofAddress(addr, 0L, loaderSession));
+        };
+    }
+
+    /**
+     * Obtains a system lookup suitable to find symbols in the standard C libraries. The set of symbols
+     * available for lookup is unspecified, as it depends on the platform and on the operating system.
+     * @return a system-specific library lookup which is suitable to find symbols in the standard C libraries.
+     */
+    static SymbolLookup systemLookup() {
+        return SystemLookup.getInstance();
+    }
+
+    /**
+     * Loads a library with given name and obtains a symbol lookup suitable to find symbols in that library.
+     * The library will be unloaded when the provided memory session is {@linkplain MemorySession#close() closed}.
+     * <p>
+     * This method is <a href="package-summary.html#restricted"><em>restricted</em></a>.
+     * Restricted methods are unsafe, and, if used incorrectly, their use might crash
+     * the JVM or, worse, silently result in memory corruption. Thus, clients should refrain from depending on
+     * restricted methods, and use safe and supported functionalities, where possible.
+     * @param name the name of the library in which symbols should be looked up.
+     * @param session the memory session which controls the library lifecycle.
+     * @return a symbol lookup suitable to find symbols in a library with given name.
+     * @throws IllegalArgumentException if {@code name} does not identify a valid library.
+     * @throws IllegalCallerException if access to this method occurs from a module {@code M} and the command line option
+     * {@code --enable-native-access} is either absent, or does not mention the module name {@code M}, or
+     * {@code ALL-UNNAMED} in case {@code M} is an unnamed module.
+     */
+    @CallerSensitive
+    static SymbolLookup libraryLookup(String name, MemorySession session) {
```

Comment:
This cannot call the path version. Note that `dlopen` has its own "known" library names. On linux I can open `libc.so.6` by name, and the correct file fill be found, using the linker cache. MacOS also has some notion of libraries that are not really files: https://developer.apple.com/forums/thread/655588

Reply:
In other words, note that the two methods are using different methods in the `RawNativeLibraries` class.

Reply:
Ah, i see. So they are not interchangeable as i thought, and the Path variant has to refer to an actual file location (relative or otherwise), where as the String might not? Might be worth clarifying that, and for the String variant it is not constrained as is the case for `System.loadLibrary`?

Reply:
> Ah, i see. So they are not interchangeable as i thought, and the Path variant has to refer to an actual file location (relative or otherwise), where as the String might not? Might be worth clarifying that, and for the String variant it is not constrained as is the case for `System.loadLibrary`?

Exactly. One checks that the path exists - the other sends the string argument straight to dlopen/LoadLibrary (which is what sometimes is required).